utils/controller_debug.py

```diff
@@ -0,0 +1,789 @@
+#   global elfFiles
+#
+#   ct = os.stat( fname ).st_ctime
+#
+#   if( not fname in elfFiles ):
+#      pass
+#   elif( elfFiles[fname]['ctime'] != ct ):
+#      pass
+#   else:
+#      return elfFiles[fname]['elf']
+#
+#   e = {}
+#   e['ctime'] = ct
+#   e['elf'] = elffile.ElfFile( fname )
+#   elfFiles[fname] = e
+#   return e['elf']
+
+
+def decodeAddr(addr, fw=None):
+
+    #   global fwname
+    #   if( fw == None ):
+    #      fw = fwname
+    #
+    if isinstance(addr, int):
+        return addr
+
+    #   if( addr[0] == '&' ):
+    #      elf = loadElf( fw )
+    #
+    #      a = elf.getSymbol( addr[1:] )
+    #      if( a != None ):
+    #         a = a.value;
+    #      return a
+    return int(addr, 0)
+
+
+def peek(addr, ct=1, fmt="+XXXX", fname=None, raw=False):
+    addr = decodeAddr(addr)
+    if addr == None:
+        print("Unknown symbol")
+        return
+
+    out = []
+
+    A = addr
+
+    while ct:
+        n = min(ct, 256)
+        dat = SendCmd(OP_PEEK, Split32(A) + Split16(n))
+        out += dat
+        ct -= len(dat)
+        A += len(dat)
+
+    if raw:
+        return out
+
+    s = FmtPeek(out, fmt, addr)
+    if fname == None:
+        print(s)
+    else:
+        fp = open(fname, "w")
+        fp.write(s)
+        fp.write("\n")
+        fp.close()
+
+
+def peek16(addr, ct=None, le=True, signed=False):
+    addr = decodeAddr(addr)
+    if addr == None:
+        print("Unknown symbol")
+        return
+
+    if ct == None:
+        ct = 1
+    out = peek(addr, 2 * ct, raw=True)
+    return Build16(out, le=True, signed=False)
+
+
+def peek32(addr, ct=None, le=True, signed=False):
+    addr = decodeAddr(addr)
+    if addr == None:
+        print("Unknown symbol")
+        return
+
+    if ct == None:
+        ct = 1
+    out = peek(addr, 4 * ct, raw=True)
+    return Build32(out, le=True, signed=False)
+
+
+# def peekf( addr, ct=None ):
+#   dat = peekl(addr,ct, le=True, signed=False);
+#   if( ct == None ):
+#      return Util.I2F(dat)
+#   ret = []
+#   for d in dat:
+#      ret.append( Util.I2F(d) )
+#   return ret
+
+
+def poke(addr, dat, ptype):
+    addr = decodeAddr(addr)
+    if addr == None:
+        print("Unknown symbol")
+        return
+    if isinstance(dat, int) or isinstance(dat, long):
+        dat = [dat]
+
+    if ptype == "long":
+        dat = Split32(dat)
+    if ptype == "short":
+        dat = Split16(dat)
+    if ptype == "float":
+        dat = [F2I(x) for x in dat]
+        dat = Split32(dat)
+
+    SendCmd(OP_POKE, Split32(addr) + dat)
+
+
+def poke32(addr, dat):
+    poke(addr, dat, "long")
+
+
+def poke16(addr, dat):
+    poke(addr, dat, "short")
+
+
+def EscCmd(buff):
+    ret = []
+    for i in buff:
+        if i == ESC or i == TERM:
+            ret.append(ESC)
+        ret.append(i)
+    ret.append(TERM)
+    return ret
+
+
+def GetResp(show=False):
+    dat = []
+    esc = False
+    if show:
+        print("Getting resp: ", end="")
+    while True:
+        x = ser.read(1)
```

Comment:
I'm not quite following what this loop does, could you add a 1-2 sentence comment?